controllers/cruisecontroltask_controller.go

```diff
@@ -70,121 +82,225 @@ func (r *CruiseControlTaskReconciler) Reconcile(ctx context.Context, request ctr
+		removeTask.SetStateScheduled()
+
+	case tasksAndStates.NumActiveTasksByOp(banzaiv1alpha1.OperationRebalance) > 0:
 		logDirsByBroker, err := scaler.LogDirsByBroker()
 		if err != nil {
 			log.Error(err, "failed to get list of volumes per broker from Cruise Control")
 			return requeueAfter(DefaultRequeueAfterTimeInSec)
 		}
-
 		brokerIDs := make([]string, 0)
-		for _, task := range tasksAndStates.GetActiveTasksByOp(OperationRebalanceDisks) {
-			if task == nil || task.IsDone() {
-				continue
-			}
+		unavailableBrokerIDs := make([]string, 0)
+		for _, task := range tasksAndStates.GetActiveTasksByOp(banzaiv1alpha1.OperationRebalance) {
+			brokerIDs = append(brokerIDs, task.BrokerID)
+			found := false
 			if onlineDirs, ok := logDirsByBroker[task.BrokerID][scale.LogDirStateOnline]; ok {
-				found := true
 				for _, dir := range onlineDirs {
-					if !strings.HasPrefix(strings.TrimSpace(dir), strings.TrimSpace(task.Volume)) {
-						found = false
+					if strings.HasPrefix(strings.TrimSpace(dir), strings.TrimSpace(task.Volume)) {
+						found = true
 					}
 				}
-				if found {
-					brokerIDs = append(brokerIDs, task.BrokerID)
+				if !found {
+					unavailableBrokerIDs = append(unavailableBrokerIDs, task.BrokerID)
 				}
-			} else {
-				task.Err = "log dir is not available in Cruise Control"
 			}
 		}
+		if len(unavailableBrokerIDs) > 0 {
+			log.Info("requeue as there are offline broker log dirs for rebalance", "brokerIDs", unavailableBrokerIDs)
+			// This requeue is not necessary because the cruiseContrlOperation controller retries the errored task
+			return requeueAfter(DefaultRequeueAfterTimeInSec)
+		}
 
-		details := []interface{}{"operation", "rebalance disks", "brokers", brokerIDs}
-
-		result, err := scaler.RebalanceDisks(brokerIDs...)
+		cruiseControlOpRef, err := r.rebalanceDisks(ctx, instance, operationTTLSecondsAfterFinished, brokerIDs)
 		if err != nil {
-			log.Error(err, "re-balancing disk(s) in Kafka cluster via Cruise Control failed", details...)
+			requeueWithError(log, fmt.Sprintf("creating CruiseControlOperation for re-balancing disks has failed, brokerIDs: %s", brokerIDs), err)
 		}
 
-		for _, task := range tasksAndStates.GetActiveTasksByOp(OperationRebalanceDisks) {
+		for _, task := range tasksAndStates.GetActiveTasksByOp(banzaiv1alpha1.OperationRebalance) {
 			if task == nil {
 				continue
 			}
-			task.FromResult(result)
+			task.SetCruiseControlOperationRef(cruiseControlOpRef)
+			task.SetStateScheduled()
 		}
 	}
 
 	if err = r.UpdateStatus(ctx, instance, tasksAndStates); err != nil {
 		log.Error(err, "failed to update Kafka Cluster status")
 	}
-	return requeueAfter(DefaultRequeueAfterTimeInSec)
+
+	return reconciled()
 }
 
-// UpdateStatus updates the Status of the provided kafkav1beta1.KafkaCluster instance with the status of the tasks
+func getBrokersAvailability(scaler scale.CruiseControlScaler, brokerIDs []string) ([]string, error) {
```

Comment:
Can we please use `getUnavailableBrokers` instead? It is a bit misleading using `getBrokersAvailability` to get the unavailable broker IDs

Reply:
Lol I just change the check to get :)